toolchain/parse/handle_var.cpp

```diff
@@ -6,22 +6,44 @@
 
 namespace Carbon::Parse {
 
-// Handles VarAs(Semicolon|For).
-static auto HandleVar(Context& context, State finish_state) -> void {
-  context.PopAndDiscardState();
+// Handles VarAs(Declaration|For).
+static auto HandleVar(Context& context, State finish_state,
+                      Lex::Token returned_token = Lex::Token::Invalid) -> void {
+  auto start_token = context.PopState().token;
 
-  // These will start at the `var`.
-  context.PushState(finish_state);
-  context.PushState(State::VarAfterPattern);
+  // These will start at the `var` or `returned`.
+  context.PushState(finish_state, start_token);
+  context.PushState(State::VarAfterPattern, start_token);
```

Comment:
We typically take the full state, not just the token (even if we don't do more). Also, this variant of PushState says "Used when forming a new subtree with a token that isn't the start of the subtree.", which isn't quite what's happening here. Would you be okay with:

```suggestion
  auto state = context.PopState();

  // These will start at the `var` or `returned`.
  state.state = finish_state;
  context.PushState(state);
  state.state = State::VarAfterPattern;
  context.PushState(state);
```

Note, I think this may also be a little more resilient with code changes, because it propagates subtree_start and has_error. The current code works because there are no token being consumed and no nodes produced between, and implicit requirement.

Though, more broadly, maybe you'd be happier with something closer to what you had previously:

```
static auto HandleVar(Context& context, State finish_state,
                      bool is_returned) -> void {
  context.PopAndDiscardState();

  // These will start at the `var` or `returned`.
  context.PushState(finish_state);
  context.PushState(State::VarAfterPattern);

  if (is_returned) {
    auto returned_token = context.ConsumeChecked(Lex::TokenKind::Returned);
    if (auto var_token = context.ConsumeIf(Lex::TokenKind::Var)) {
      context.AddLeafNode(NodeKind::VariableIntroducer, *var_token);
      context.AddLeafNode(NodeKind::ReturnedSpecifier, returned_token);
    } else {
      CARBON_DIAGNOSTIC(ExpectedVarAfterReturned, Error,
                        "Expected `var` after `returned`.");
      context.emitter().Emit(*context.position(), ExpectedVarAfterReturned);
      auto semi = context.SkipPastLikelyEnd(returned_token);
      context.AddLeafNode(NodeKind::EmptyDeclaration,
                          semi ? *semi : returned_token,
                          /*has_error=*/true);
      return;
    }
  } else {
    context.AddLeafNode(NodeKind::VariableIntroducer, context.ConsumeChecked(Lex::TokenKind::Var));
  }

  // This will start at the pattern.
  context.PushState(State::PatternAsVariable);
}
```

Reply:
I know we do this in other places already, but I find this idiom of "mutate an existing state and then push it" to be confusing. It's very unclear to me what that means -- which parts of the existing state are being intentionally preserved, and why -- and it's unclear that it would remain correct as more information is added to the state. (For example, this propagates the `has_error` bit from the parent state to two of the three child states, but is that really intentional or just happenstance?)

I'm making this change anyway, but I think we should think a bit about whether there's a way to make this kind of state manipulation clearer. Perhaps there's a different way that we can think about states that would help, such as, as a pending parse node.

(time passes)

I ended up using this modify-and-push approach for the first state only. That's the one that represents a pending parse node, and really should be inheriting information from the parent state. The second state doesn't represent a pending parse node, and its token and subtree start don't matter, so I think it's simpler to push a fresh state in that case. This also, I think, makes it clearer that the `has_error` flag is inherited onto only a single node.

> The current code works because there are no token being consumed and no nodes produced between, and implicit requirement.

I think it only requires that no nodes are produced between -- tokens *were* being consumed here, which is why we needed to pass the original start token to `PushState`. But yeah, it's definitely brittle -- and in fact I think maybe it's even more brittle than you were suggesting, because some changes work and some don't.